### PR TITLE
DEV-38: unchecked collateral withdraw

### PR DESCRIPTION
**Pull request by zbirenbaum**
Implements DEV-38 - unchecked collateral withdraw

A couple things left to do here so marked it as a draft, but wanted to get the core implementation up so it could be discussed.

I think we need an Interface for contracts which implement `onUncheckedCollateralWithdraw`

Is the AMM helper the only contract planned to use this, or should there be a IUncheckedCollateralHelper which the AMM helper inherits from?

Right now I call the function directly using low level calls, but standardizing this would probably be best.

Sorry in advance if this didn't use the linear integrations properly, I was having a little trouble with it.

### CODE REVIEW

File: src/contracts/minter/facets/DepositWithdrawFacet.sol [hunk 1]
```diff
@@ -87,4 +87,81 @@ contract DepositWithdrawFacet is DiamondModifiers, MinterModifiers, IDepositWith
...
+        IERC20Upgradeable(_collateralAsset).safeTransfer(msg.sender, _withdrawAmount);
+
+        // execute the callback on the caller
+        _postUncheckedCollateralWithdrawal(_account, _collateralAsset, _withdrawAmount, _depositedCollateralAssetIndex);
+
+        /*
+         Perform the MCR check after the callback has been executed
+         Ensures accountCollateralValue remains over accountMinColateralValueAtRatio(MCR)
+        */
+        ms().verifyAccountCollateral(_account, _collateralAsset, _withdrawAmount);
```
Comment (panukettu) [inline src/contracts/minter/facets/DepositWithdrawFacet.sol:138]: `_withdrawAmount` should be 0 here as we already reduce deposits in `recordCollateralWithdrawal`

File: src/contracts/minter/facets/DepositWithdrawFacet.sol [hunk 2]
```diff
@@ -87,4 +87,81 @@ contract DepositWithdrawFacet is DiamondModifiers, MinterModifiers, IDepositWith
...
+        uint _withdrawAmount,
+        uint _depositedCollateralAssetIndex
+    ) internal returns (bytes memory) {
+        /*
+         Executes the callback on the caller after sending them the withdrawn collateral
+         - Not sure where to put the AMM helper interface so temporarily using to execute the CB
+            TODO: Make interface representing contracts that implement onUncheckedWithdraw instead of using low level call
+        */
+        string memory callbackSignature = "onUncheckedCollateralWithdraw(address,address,uint256,uint256)";
+        (bool success, bytes memory data) = Meta.msgSender().call(
```
[the remark this answers is not in the document]
Comment (panukettu) [reply src/contracts/minter/facets/DepositWithdrawFacet.sol:155]: will be much simpler if we just inline this after the interface exists

**Comment by panukettu** [comment]
> Implements DEV-38 - unchecked collateral withdraw
> 
> A couple things left to do here so marked it as a draft, but wanted to get the core implementation up so it could be discussed.
> 
> I think we need an Interface for contracts which implement `onUncheckedCollateralWithdraw`
> 
> Is the AMM helper the only contract planned to use this, or should there be a IUncheckedCollateralHelper which the AMM helper inherits from?
> 
> Right now I call the function directly using low level calls, but standardizing this would probably be best.
> 
> Sorry in advance if this didn't use the linear integrations properly, I was having a little trouble with it.

1. Yes lets create a simple interface for a receiver eg. `ICollateralReceiver`. We could use `ERC165.supportsInterface` to ensure this on the function but not sure if that is overdoing it. 

2. This is generic. Contracts using this will use `IKresko.withdrawCollateralUnchecked` and implement `ICollateralReceiver.onUncheckedColalteralWithdraw`. No need for other considerations.

**Comment by zbirenbaum** [comment]
I resolved all the issues in the comments, this should be ready for final review.

**Comment by tabshaikh** [comment]
would love to see some tests the function

File: testbroke.sh [hunk 3]
```diff
@@ -0,0 +1 @@
+hh test src/test/minter/00-init.ts src/test/minter/01-configuration.ts src/test/oracle/* --deploy-fixture
```
Comment (tabshaikh) [inline testbroke.sh:1]: maybe this whole file can be removed

Comment (zbirenbaum) [reply testbroke.sh]: `git add .` strikes again

**Comment by zbirenbaum** [comment]
> would love to see some tests for this change

I thought about that too but was thinking they might belong thematically in the tests for something implementing ICollateralReceiver like the AMM Helper to make sure it can't withdraw too much, and that after it does something with the collateral that the user's MCR check still passes. Implementing a contract that can properly make use of this function would require actually doing something with the user's collateral.

**Review by panukettu — CHANGES_REQUESTED** [review]
It is looking good overall but lets scope this down a bit: Drop access without `Role.MANAGER` and remove mutex.  
We need to demonstrate few scenarios with tests as @tabshaikh suggested. Lets demonstrate following (at least):

1. Can do nothing after withdrawing some collateral in the receiver.
2. Can re-deposit withdrawn collateral back into the protocol.
3. Can deposit another collateral into the protocol for `_account` from the receiver.
4. Reverts because account ends up below MCR after callback.
5. Reverts because withdrawal amount is greater than account deposits. 
6. Reverts because caller is not `Role.MANAGER`.

**Comment by zbirenbaum** [comment]
> * Can do nothing after withdrawing some collateral in the receiver if above MCR after callback.
> * Can re-deposit withdrawn collateral back into the protocol for `_account`.
> * Can deposit another collateral into the protocol for `_account`.
> * Reverts because account ends up below MCR after callback.
> * Reverts because withdrawal amount is greater than account deposits.
> * Reverts because caller is not `Role.MANAGER`.

I got a little stuck on trying to implement a `MockCollateralReceiver` contract. I looked at the other facet contracts for a while but still couldn't quite figure out exactly what it should be inheriting. It doesn't look like other facets have to be initialized but I'm not sure how I should be resolving the address of the Diamond proxy. Could you point me in the right direction or link contract(s) that require the some or all of the same basic protocol information as what is required here?

**Comment by panukettu** [comment]
LGTM!

**Pull request merged**